src/oidc/tokenRefresher.ts

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { IdTokenClaims, OidcClient, WebStorageStateStore } from "oidc-client-ts";
+
+import { TokenRefreshFunction } from "../http-api";
+import { IDelegatedAuthConfig } from "../client";
+import { generateScope } from "./authorize";
+import { discoverAndValidateAuthenticationConfig } from "./discovery";
+import { logger } from "../logger";
+
+/**
+ * @experimental
+ * Class responsible to refreshing OIDC access tokens
+ */
+export abstract class OidcTokenRefresher {
+    public readonly oidcClientReady!: Promise<void>;
```

Comment:
I'm not entirely clear how this is meant to be used. I can only find references to it in the tests - if that's the only place then it's probably better to make it private and put some `ts-ignore` annotations in the tests?

If it's going to be used elsewhere, could you document it?